presto-native-execution/presto_cpp/main/http/HttpClient.cpp

```diff
@@ -352,6 +353,138 @@ class ConnectionHandler : public proxygen::HTTPConnector::Callback {
   std::unique_ptr<proxygen::HTTPConnector> connector_;
 };
 
+namespace {
+// Same value as
+// https://github.com/prestodb/presto/blob/831d5947b909fee0d5c0091a3246ddc5b31b2731/presto-main/src/main/java/com/facebook/presto/server/ServerMainModule.java#L547
+constexpr int kMaxConnectionsPerServer = 250;
+} // namespace
+
+std::pair<proxygen::SessionPool*, proxygen::ServerIdleSessionController*>
+HttpClientConnectionPool::getSessionPoolImpl(SessionPools& endpointPools) {
+  auto* evb = folly::EventBaseManager::get()->getExistingEventBase();
+  VELOX_CHECK_NOT_NULL(evb);
+  {
+    auto rlock = endpointPools.byEventBase.rlock();
+    auto it = rlock->find(evb);
+    if (it != rlock->end()) {
+      return {it->second.get(), &endpointPools.idleSessions};
+    }
+  }
+  // NOTE: local event base is unique so only 1 thread can access the same key
+  // entry.
+  auto wlock = endpointPools.byEventBase.wlock();
+  auto& pool = (*wlock)[evb];
+  VELOX_CHECK_NULL(pool);
+  pool = std::make_unique<proxygen::SessionPool>(
+      nullptr,
+      kMaxConnectionsPerServer,
+      std::chrono::seconds(30),
+      std::chrono::milliseconds(0),
+      nullptr,
+      &endpointPools.idleSessions);
+  return {pool.get(), &endpointPools.idleSessions};
+}
+
+std::pair<proxygen::SessionPool*, proxygen::ServerIdleSessionController*>
+HttpClientConnectionPool::getSessionPool(const proxygen::Endpoint& endpoint) {
+  {
+    auto rlock = pools_.rlock();
+    auto it = rlock->find(endpoint);
+    if (it != rlock->end()) {
+      return getSessionPoolImpl(*it->second);
+    }
+  }
+  auto wlock = pools_.wlock();
+  auto it = wlock->find(endpoint);
+  if (it != wlock->end()) {
+    auto rlock = wlock.moveFromWriteToRead();
+    return getSessionPoolImpl(*it->second);
+  }
+  auto& endpointPools = (*wlock)[endpoint];
```

Comment:
nit:
```
auto endpointPools = std::make_unique<SessionPools>();
endpointPools->idleSessions.setMaxIdleCount(kMaxConnectionsPerServer);
(*wlock)[endpoint] = std::move(endpointPools);
```

Reply:
We still need the `endpointPools` later so cannot move it